aiida_vasp/assistant/parameters.py

```diff
@@ -528,3 +541,24 @@ def inherit_and_merge_parameters(inputs):
     update_nested_dict(parameters, input_parameters)
 
     return parameters
+
+
+def vasp_parameter_nesting(inputs, namespaces):
+    """Helper function to make sure that the namespaces are properly handled when they are nested."""
+    try:
+        # inputs might not have parameters, or parameters might be empty
+        if 'vasp' in inputs.parameters.get_dict():
```

Comment:
But would you also not then remove the possibility of the something that tries to modify the inputs actually being successful. This might also be something you would want? I think, it possible we should try not to introduce vasp functionality (or namespace) in the higher workchains (unless the user wants override and specifies that directly in the `parameters` and then it flows through). Or?

Reply:
I thought that the idea was to ensure that there was a `vasp` namespace to separate all the `INCAR` like inputs from the rest of the data that can be passed in `parameters`, maybe I misunderstood. 

The question is then if we do not separate the `INCAR` parameters from the rest of the other parameters, we would need to set an exception for parameters such as the ones belonging from `dynamics`.  Since if we do not separate them, they would be read as invalid keys.

Reply:
We should have a separate `vasp` namespace in which the `INCAR` tags are stored, so that when `VaspWorkChain` receives `parameters` we have e.g. `parameters.vasp` and `parameters.dynamic` or something like that.

Reply:
Aha, I think (unless I completely did things wrong) that is what is being done here, the `parameters.vap` is supposed to contain only the `INCAR` parameters.

Reply:
@JPchico Okey. I do not think we need this test. As far as I see it, we should maybe allow parameters in both the `parameters.vasp` and `parameters.dynamics` (etc.) namespace at this point and this test removes that option. As soon as something is already set in `parameters.vasp`, say by the user, anything else in `parameters.namespace` is not included from `parameters`. So maybe we can just remove it and add the `vasp` namespace explicitly here so that it is always included?

Reply:
> @JPchico Okey. We should remove all VASP traces at this point as the workchain should be able to run on QE (as an example). We execute this code in the workchains, so they need to be code agnostic. We have two options:
> 

@espenfl sure thing I can see about removing it, I understand what you say about the `vasp` name. It is not ideal to have something hardcoded as `vasp` if one wants to share the utilities with other codes.

> * Rename `vasp` to something else, that is code agnostic. The working principle should be very similar for the other codes and the `ParameterMassager` can handle adjustments. Maybe `override` or similar?

This could be, one could call it `code` maybe? as these are very code specific parameters? or something like that.

> * Remove tests on namespaces (retain original code). That would allow namespaces that are not allowed to pass and a user might be confused, so maybe we should consider to add a test like you did and also eject an error if a namespace that we do not allow is detected. But then we should rename the `vasp` namespace.

This could be, I need to try to replicate the original error I was getting to see exactly how to get around this.

Reply:
> This could be, one could call it `code` maybe? as these are very code specific parameters? or something like that.

@JPchico Okey, let us go with `code` for now. I do not think users will mix this up with the AiiDA code so this should be fine and is also contextual. Thanks for the suggestion.

> This could be, I need to try to replicate the original error I was getting to see exactly how to get around this.

I think if we go with `code` we should keep the tests for the namespace in `parameters` that you added as well, but remove the test. Maybe we should also raise a warning if we detect a namespace that is not in the list. Now it is silently ignored, which might confuse users. Great if you could manage to pinpoint why you needed to add this specifically.

Reply:
@espenfl So I finally had some time to check this. I think that the reason behind me adding this was the test `test_inherit_and_merge` in `test_parameters.py`. What seems to happen here is that when just using the function `inherit_and_merge_parameters` in an `input` that does not have a `code` (previous `vasp`) entry will result in an output without the `code` key. This is then compared with a dictionary that has the `code` key. Which of course results in an assertion error.

That is why I added the check, since in case you do not add any keys that would belong to the `code` entry no `code` entry is created, this would cause problems later on when the workchains will start to ask for the presence of the `code` key.

Perhaps this can be handled in a more elegant fashion, but this is why I added this entry.

Reply:
@JPchico Sounds great. Thanks for explaining and investigating. Maybe before pushing the PR you can add a small comment regarding this in the code. E.g. that we need to establish the `code` namespace if it does not already exist.